### PR TITLE
Add wms window

### LINKED ISSUE #250: Integerate BasiGX AddWMS Form
Add a new button to the layer tree that allows a user to open the [AddWms](http://terrestris.github.io/BasiGX/master/docs/#!/api/BasiGX.view.form.AddWms) form and add in external WMS layers. 
These would then be added dynamically to the layer tree (in a new "External Layers" folder) for the session. 

Is there a demo of this feature available online?

Some questions:

- can a default list of WMS servers be provided (I believe this could be done using the wmsBaseUrls property)?
- are there any plans to preserve added layers across sessions using local storage?

### PR DESCRIPTION
**Pull request by simonseyock**
This resolves #250 

While doing this yesterday I moved some code around. This might be confusing for others, I can undo this. I created a new controller for the layertree and moved the model manipulation from `view/LayerTree.js` into `controller/LayerTreeController.js`.

In my opinion this code belongs into a controller instead of the view.

### CODE REVIEW

**Comment by simonseyock** [comment]
This is how it looks:

Peek 2020-09-29 16-47

**Comment by geographika** [comment]
@simonseyock - thanks for this. Will be giving a test this morning and get back to you.

**Comment by simonseyock** [comment]
I will move the refactoring into a second PR. Is a force push on this branch ok"

File: app/view/LayerTree.js
Comment (geographika) [inline app/view/LayerTree.js:38]: Change text to "Add External Map Layer"

**Comment by geographika** [comment]
Thanks @simonseyock - I have this integrated and working in one of full applications. 

A few questions:

- How do we configure the button text in the form to be in English? Does this require updates to BasiGX?
- How is the WMS URL configured (currently set to https://ows.terrestris.de/osm/service)? Is it possible to have a dropdown box with some preset servers plus allow free text entry?
- Is it possible to add in the layer "Refresh" and "Opacity" layer plugins to the added layers?

image

For my own reference (but please clarify if incorrect), to add to an application based on cpsi-mapview:

- Add a reference to `CpsiMapview.view.addWms.AddWmsWindow` to the application Main view 
- Add a new `xtype: 'cmv_add_wms_window'` to the main view. This is the form that allows a user to add a WMS. It is hidden by default, and reused when a user clicks on the button in the layer tree

When the form is created it adds a listener to the button click event to be displayed. 
When the first layer is added a new group layer is added to the map (and hence the layer tree).

**Comment by simonseyock** [comment]
I will look into your comments. Is a force push on this branch ok?

File: app/view/addWms/AddWmsWindow.js
Comment (simonseyock) [reply app/view/addWms/AddWmsWindow.js:4]: Yes this is no problem.

File: app/controller/LayerTreeController.js
Comment (simonseyock) [reply app/controller/LayerTreeController.js:5]: I add these changes to refactor PR

**Comment by simonseyock** [comment]
> How do we configure the button text in the form to be in English? Does this require updates to BasiGX?
The translations can be added by changing the viewModel of the `view.addWms.AddWmsForm`.

I added some translations.

> How is the WMS URL configured (currently set to https://ows.terrestris.de/osm/service)? Is it possible to have a dropdown box with some preset servers plus allow free text entry?

This can be done easily with the property `wmsBaseUrls` on the `view.addWms.AddWmsForm` like this:
```
    wmsBaseUrls: [
        'https://ows.terrestris.de/osm/service',
        'url2'
    ],
```
It automatically shows a combo box and text input is still allowed.

> Is it possible to add in the layer "Refresh" and "Opacity" layer plugins to the added layers?

I added the options to the layers.

**Comment by simonseyock** [comment]
> For my own reference (but please clarify if incorrect), to add to an application based on cpsi-mapview:
>
>    Add a reference to CpsiMapview.view.addWms.AddWmsWindow to the application Main view
>    Add a new xtype: 'cmv_add_wms_window' to the main view. This is the form that allows a user to add a WMS. It is hidden by default, and reused when a user clicks on the button in the layer tree
>
> When the form is created it adds a listener to the button click event to be displayed.
> When the first layer is added a new group layer is added to the map (and hence the layer tree).

This is correct. To change the name of the layerGroup the property `layerGroupName` of the `view.addWms.AddWmsForm` can be changed.

**Comment by geographika** [comment]
> 
> 
> I will look into your comments. Is a force push on this branch ok?

Yes fine with me.

**Comment by simonseyock** [comment]
I added the requested changes. This branch is now rebased on layer-tree-refactoring.

**Comment by geographika** [comment]
Thanks @simonseyock. I'm happy to merge this. 

For reference - this is how I'm adding and configuring in an application:

```js
         items: [
...
// layer tree declaration
...
        {
            xtype: 'cmv_add_wms_window',
            title: 'Add External WMS Layer',
            items: [
                {
                    xtype: 'cmv_add_wms_form',
                    wmsBaseUrls: [
                        'https://ows.terrestris.de/osm/service',
                        'url2'
                    ]
                }
            ]
          }
...
        ]
```

One thing to watch out for is that the form has to be added to the viewport's items **after** the layertree component declaration (I'm adding it outside the layertree otherwise it is constrained to the layertree panel when moving the window). 
Otherwise, the following line in `setupListener` returns an empty list as the layertree component is not created:

`var layerTrees = Ext.ComponentQuery.query('cmv_layertree');`

**Review by geographika — APPROVED** [review]

File: app/controller/addWms/AddWmsFormController.js
Comment (chrismayer) [inline app/controller/addWms/AddWmsFormController.js:6]: Please add some API docs with a short description here.

File: app/controller/addWms/AddWmsWindowController.js
Comment (chrismayer) [inline app/controller/addWms/AddWmsWindowController.js:10]: Please add some API docs with a short description here.

File: app/view/LayerTree.js
Comment (chrismayer) [inline app/view/LayerTree.js:34]: Something like 

```
bbar: [{
  text: 'Add External Map Layer',
...
}]
```
should do the same and would make 

```
 header: {
        dock: 'bottom'
    },
```
obsolete.

Just for information, no need to change.

File: app/view/addWms/AddWmsForm.js
Comment (chrismayer) [inline app/view/addWms/AddWmsForm.js:21]:
```suggestion
            requestLayersBtnText: 'Request available layers',
```

File: app/view/addWms/AddWmsForm.js
Comment (chrismayer) [inline app/view/addWms/AddWmsForm.js:29]:
```suggestion
            msgServiceException: 'An OGC ServiceException has been occurred',
```

File: app/view/addWms/AddWmsForm.js
Comment (chrismayer) [inline app/view/addWms/AddWmsForm.js:39]:
```suggestion
            msgServerError: 'An unspecified server error occurred.',
```

File: app/view/main/Main.js
Comment (chrismayer) [inline app/view/main/Main.js:49]: How about instantiating the `cmv_add_wms_window` in the handler of the button instead of hanging it in the main view?

**Review by chrismayer — COMMENTED** [review]
Very cool @simonseyock. I added some minor issues you might want to have a look at before merging. TIA!

File: app/view/addWms/AddWmsForm.js
Comment (marcjansen) [inline app/view/addWms/AddWmsForm.js:29]:
```suggestion
            msgServiceException: 'An OGC ServiceException has occurred',
```

**Comment by simonseyock** [comment]
I added the suggestions. The window is now created inside the LayerTreeController like this:
```
            this.addWmsWindow = Ext.create('CpsiMapview.view.window.MinimizableWindow', {
                items: [
                    {
                        xtype: 'cmv_add_wms_form'
                    }
                ],
            });
            this.addWmsWindow.show();
```
This way it does not need to be part of the main view.

**Review by chrismayer — APPROVED** [review]

**Comment by geographika** [comment]
@simonseyock - there is a conflict to fix due to updates to the LayerTreeController. After those are fixed I think this is fine to merge - thoughts @chrismayer ?

**Comment by chrismayer** [comment]
Yes, this is good to go (after resolving).

**Comment by simonseyock** [comment]
@geographika I resolved the merge conflicts

**Pull request merged**
